rs/embedders/src/serialized_module.rs

```diff
@@ -102,3 +111,305 @@ impl SerializedModule {
+            .expect("Unable to serialize module: failed to create initial state file");
+        initial_state_file
+            .write_all(
+                &bincode::serialize(&initial_state_data)
+                    .expect("Unable to serialize module: failed to serialize initial state"),
+            )
+            .expect("Unable to serialize module: failed to write initial state file");
+
+        // Set file permissions to readonly and reopen with new permissions.
+        let mut permissions = bytes_file
+            .metadata()
+            .expect("Unable to serialize module: failed to get bytes file permissions")
+            .permissions();
+        permissions.set_readonly(true);
+        bytes_file
+            .set_permissions(permissions.clone())
+            .expect("Unable to serialize module: failed to set bytes file permissions");
+        initial_state_file
+            .set_permissions(permissions)
+            .expect("Unable to serialize module: failed to set initial state file permissions");
+        let bytes_file = File::open(bytes_path)
+            .expect("Unable to serialize module: failed to reopen bytes file");
+        let initial_state_file = File::open(initial_state_path)
+            .expect("Unable to serialize module: failed to reopen initial state file");
+
+        // Delete the files so that they are automatically cleaned up
+        // when there are no more descriptors.
+        std::fs::remove_file(bytes_path)
+            .expect("Unable to serialize module: failed to delete bytes file");
+        std::fs::remove_file(initial_state_path)
+            .expect("Unable to serialize module: failed to delete initial state file");
+
+        Self {
+            bytes: bytes_file,
+            initial_state_data: initial_state_file,
+            compilation_cost: serialized_module.compilation_cost,
+            imports_details: serialized_module.imports_details,
+            is_wasm64: serialized_module.is_wasm64,
+        }
+    }
+
+    /// Map the initial state file and deserialize its contents.
+    pub(crate) fn initial_state_data(&self) -> InitialStateData {
+        // Mmap the initial state file so that the file descriptor isn't mutated
+        // (they might be shared).
+        let mmap_size = self
+            .initial_state_data
+            .metadata()
+            .expect("Unable to get size for initial state data file")
+            .size() as usize;
+        // Safety: Rust guarantees that the fd is valid and the null pointer
+        // argument implies that this won't mess with any existing memory.
+        let mmap_ptr = unsafe {
+            mmap(
+                std::ptr::null_mut(),
+                mmap_size,
+                ProtFlags::PROT_READ,
+                MapFlags::MAP_PRIVATE,
+                self.initial_state_data.as_raw_fd(),
+                0,
+            )
+        }
+        .unwrap_or_else(|err| {
+            panic!(
```

Comment:
Can I assume that all new panics and unwraps are covered by the docstring above? I.e., we are not doing worse than before, because allocating memory could always fail. 

It seems to me that we have more interactions that might fail now, but the source of failures is now not memory allocation but the filesystem. I suppose that does not matter in this context, only that we cannot recover from any of these, so panicking is the right thing to do, correct? 

What is the context that fails on a panic? All of replica? Or is this an isolated task that fails? 

Is there a component in replica that knows about available system resources (disk, memory, file handles, ...) and can choose not to create new sandbox processes, or is this choice only governed by a limit on the number of processes that we control manually?

Reply:
Yeah, I basically structured this so that any fs error is treated the same as a memory allocation failure which causes a panic. The effect of this will be that the node restarts (with a cleared compilation cache and no running sandboxes). So it's possible that as a whole the subnet could recover, but it could also end up in a crash loop.

We could take into account system resources and fall back to other logic in that case, but I'm not sure if it would be that helpful. Hopefully we can bound the disk space of the cache to something that's more than we should need without risking running out of space.

Reply:
Thanks for the context!

>the node restarts 

The replica process, or the GuestOS, or even deeper down?